Source/WebCore/Modules/cookie-store/CookieStore.cpp

```diff
@@ -314,11 +318,93 @@ void CookieStore::remove(CookieStoreDeleteOptions&& options, Ref<DeferredPromise
     set(WTFMove(initOptions), WTFMove(promise));
 }
 
+void CookieStore::cookiesAdded(const String& host, const Vector<Cookie>& cookies)
+{
+    UNUSED_PARAM(host);
+    if (!m_hasChangeEventListener)
+        return;
+
+    auto* context = scriptExecutionContext();
+    if (!context)
+        return;
+
+    auto& document = *downcast<Document>(context);
+    UNUSED_PARAM(document);
+    ASSERT(host == document.url().host().toString());
+
+    Vector<CookieListItem> changedCookieList;
+    Vector<CookieListItem> deletedCookieList;
+    auto currentTime = WallTime::now().secondsSinceEpoch().milliseconds();
+    for (auto cookie : cookies) {
+        if (cookie.expires && *cookie.expires < currentTime) {
+            cookie.value = nullString();
+            deletedCookieList.append(CookieListItem { cookie });
+        } else
+            changedCookieList.append(CookieListItem { cookie });
+    }
+
+    CookieChangeEventInit eventInit;
+    eventInit.changed = WTFMove(changedCookieList);
+    eventInit.deleted = WTFMove(deletedCookieList);
+    queueTaskToDispatchEvent(*this, TaskSource::DOMManipulation, CookieChangeEvent::create(eventNames().changeEvent, WTFMove(eventInit), CookieChangeEvent::IsTrusted::Yes));
+}
+
+void CookieStore::cookiesDeleted(const String& host, const Vector<Cookie>& cookies)
+{
+    UNUSED_PARAM(host);
+    if (!m_hasChangeEventListener)
+        return;
+
+    auto* context = scriptExecutionContext();
+    if (!context)
+        return;
+
+    auto& document = *downcast<Document>(context);
+    UNUSED_PARAM(document);
+    ASSERT(host == document.url().host().toString());
+
+    auto cookieList = cookies.map([](auto cookie) {
+        cookie.value = nullString();
+        return CookieListItem { cookie };
+    });
+
+    CookieChangeEventInit eventInit;
+    eventInit.deleted = WTFMove(cookieList);
+    queueTaskToDispatchEvent(*this, TaskSource::DOMManipulation, CookieChangeEvent::create(eventNames().changeEvent, WTFMove(eventInit), CookieChangeEvent::IsTrusted::Yes));
+}
+
 const char* CookieStore::activeDOMObjectName() const
 {
     return "CookieStore";
 }
 
+void CookieStore::stop()
+{
+    if (!m_hasChangeEventListener)
+        return;
+
+    auto* context = scriptExecutionContext();
+    if (!context)
```

Comment:
Hmm, Actually I believe this is always be null when `stop()` gets called I believe. I think this shows that we should store both the `CookieJar` and the `host` as data members.